### PR TITLE
[TR] PIM-2735: Introduce the modal-form component

### PR DESCRIPTION
**Pull request by grena**
| Q                 | A
| ----------------- | ---
| Added Specs       | N
| Added Behats      | N
| Travis CI is ok   | Y
| Changelog updated | N

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+             */
+            formName: '',
+
+            /**
+             * You must implement this method to return the serialized custom form data.
+             *
+             * @return {Object}
+             */
+            getData: function () {
+                return {};
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js]: throw an exception

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+            /**
+             * @return {object}
+             */
+            getModalParameters: function () {
+                return {
+                    allowCancel: true,
+                    okCloses:    false,
+                    content:     '',
+                    title:       _.__('pimee_enrich.entity.product_draft.modal.send_for_approval'),
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js]: nope, not in community

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+ * The custom form must be passed in as a service.
+ *
+ * A deferred object is returned on rendering:
+ * - Success: Resolved when the user clicks on the OK button, the callback contains the form data.
+ * - Fail: the user canceled the modal form.
+ *
+ * Typical use example
+ * ===================
+ *
+ * var myModalForm = new ModalForm();
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js]: FormModal

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * myModalForm.render()
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js]: open

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        'pim/form-builder'
+    ],
+    function (
+        $,
+        _,
+        Backbone,
+        BaseForm,
+        FormBuilder
+    ) {
+        return BaseForm.extend({
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js]: Backbone.view.extend(

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js (hunk 1 above)
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js]: this.form.getData()

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+            },
+
+            /**
+             * Render the modal with the custom form service.
+             * Returns the deferred object to catch success (OK) & fail (Cancel) event of the modal.
+             *
+             * @returns {Deferred}
+             */
+            render: function () {
+                var deferred = $.Deferred();
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/modal-form.js]: remove

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
+'use strict';
+
+/**
+ * This service instantiates a modal with a custom form.
+ * The custom form must be passed in as a service.
+ *
+ * A deferred object is returned on rendering:
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: on modal opening

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+ *
+ * var myFormModal = new FormModal('pim-product-edit-form');
+ *
+ * myFormModal.open()
+ *      .then(function(myFormData) {
+ *          // user filled the form and clicked OK, now handle data
+ *      })
+ *      .fail(function() {
+ *          // user clicked Cancel button
+ *      });
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: :heart:

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        'pim/form-builder'
+    ],
+    function (
+        $,
+        _,
+        Backbone,
+        FormBuilder
+    ) {
+        return Backbone.View.extend({
+
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: useless blank line

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+                cancelText:  '[cancel]',
+                modalOptions: {
+                    backdrop: 'static',
+                    keyboard: false
+                }
+            },
+
+            /**
+             * @param {string} formName
+             * @param {object} modalParameters
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: `Object` takes a capital letter (same in all the PR)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+             *
+             * @return {Deferred}
+             */
+            open: function () {
+                var deferred = $.Deferred();
+
+                FormBuilder
+                    .build(this.formName)
+                    .then(function (form) {
+                        var modal = new Backbone.BootstrapModal(this.getModalParameters());
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: this.modal

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+                        modal.open();
+
+                        form.setElement(modal.$('.modal-body')).render();
+
+                        modal.on('cancel', deferred.reject);
+                        modal.on('ok', function () {
+                            var data = form.getData();
+                            deferred.resolve(data);
+                            modal.close();
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: you should not call it here and you should create a close method on this class

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+ * - Fail: the user canceled the modal form.
+ *
+ * Typical use example
+ * ===================
+ *
+ * var myFormModal = new FormModal('pim-product-edit-form');
+ *
+ * myFormModal.open()
+ *      .then(function(myFormData) {
+ *          // user filled the form and clicked OK, now handle data
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: myFormModal.close();

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+
+                FormBuilder
+                    .build(this.formName)
+                    .then(function (form) {
+                        var modal = new Backbone.BootstrapModal(this.getModalParameters());
+                        modal.open();
+
+                        form.setElement(modal.$('.modal-body')).render();
+
+                        modal.on('cancel', deferred.reject);
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: Is the modal closed when you cancel ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js (hunk 12 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: +1

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js (hunk 13 above)
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: done

File: src/Pim/Bundle/NotificationBundle/Resources/public/js/form/base-comment.js [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+            events: {
+                'change textarea': 'updateModel'
+            },
+            initialize: function () {
+                this.model = new Backbone.Model();
+
+                BaseForm.prototype.initialize.apply(this, arguments);
+            },
+            updateModel: function () {
+                this.model.set('comment', this.$('textarea[name="' + _.__('pim_notification.comment.modal.label') + '"]').val());
```
Comment (grena) [inline src/Pim/Bundle/NotificationBundle/Resources/public/js/form/base-comment.js]: As in the model, the textarea name shouldn't be dynamic imho.
You can hard code textarea name to "comment".
Plus this line is too long! :P

File: src/Pim/Bundle/NotificationBundle/Resources/public/templates/comment/comment.html [hunk 16]
```diff
@@ -1,4 +1,4 @@
 <div class="notification-comment">
-    <div class="title"><%- title %></div>
-    <textarea name="<%- title %>"></textarea>
+    <label class="title" for="modal-comment"><%- title %></label>
```
Comment (grena) [inline src/Pim/Bundle/NotificationBundle/Resources/public/templates/comment/comment.html]: update title to "label" please

File: src/Pim/Bundle/NotificationBundle/Resources/public/js/form/base-comment.js [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+            initialize: function () {
+                this.model = new Backbone.Model();
+
+                BaseForm.prototype.initialize.apply(this, arguments);
+            },
+            updateModel: function () {
+                this.model.set('comment', this.$('textarea[id="modal-comment"]').val());
+            },
+            render: function () {
+                this.$el.html(this.template({title: _.__('pim_notification.comment.modal.label')}));
```
Comment (grena) [inline src/Pim/Bundle/NotificationBundle/Resources/public/js/form/base-comment.js]: `{label: `

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+                modalOptions: {
+                    backdrop: 'static',
+                    keyboard: false
+                }
+            },
+
+            /**
+             * @param {string} formName
+             * @param {function} submitCallback
+             * @param {Object} modalParameters
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: align please

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+                this.formName        = formName;
+                this.submitCallback  = submitCallback;
+                this.modalParameters = _.extend(this.modalParameters, modalParameters);
+            },
+
+            /**
+             * Render the modal with the custom form service.
+             * Returns the deferred object to catch success (OK) & fail (Cancel) event of the modal.
+             *
+             * @return {Deferred}
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: I think we use `{Promise}` elsewhere in the code

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+            /**
+             * The form name the modal should display.
+             * This service must be registered with RequireJS, eg: 'pim-product-edit-form'
+             */
+            formName: '',
+
+            /**
+             * Instance of the UI modal element.
+             */
+            modal: undefined,
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: null is better

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+
+            /**
+             * Callback triggered on form submission.
+             * This callback should return a promise, resolved when data validation check is OK.
+             */
+            submitCallback: undefined,
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/form-modal.js]: same here

**Comment by phaseinducer** [comment]
GTM

**Comment by willy-ahva** [comment]
GTM after squash

**Pull request merged**